### DISCUSSION TITLE
Callbacks or Adapaters or both? Working with an already external backend in a different language

### DISCUSSION THREAD

**Discussion by Oyelowo** [nextauthjs/next-auth#4230] [Help]
Hi! Great work on this library! It's been nice working with it! 👍 
One challenge I have encountered is working with an already existing external backend(in a different language) running in a Kubernetes cluster. Hence, I am using Redis to manage cookie sessions.

# Problem
I have an external `graphQL backend` written in `Rust` and already implemented `session` with `cookie` generation on the backend. I also have `JWT` implemented as an alternative(just for comparison).
I would love to reuse this implementation for my NextJS frontend. So, `JWT` or `Session` implementation
from `Next-Auth` can probably be skipped? I am using `react-query` to fetch the `user/session` data directly from the `rust graphQL backend`  and the `cookie` is automatically set by the backend.

## I am considering either of the two options :
1. Create a custom `adapter`(https://next-auth.js.org/tutorials/creating-a-database-adapter) with all the functions calling my `rust graphQL` backend to persist the data.

2.  Use the `callbacks`(https://next-auth.js.org/configuration/callbacks) and on `sign in`, create logic that stores
user's `Oauth` data in my database through my external `graphQL` backend API and generates a new session cookie in the process.

```ts
...
  callbacks: {
    async signIn({ user, account, profile, email, credentials }) {
    # This should create or update user data and generate a new cookie session for the user
       const user = callToRustGraphqlBackendForUserData("github", {userData})
      return true
    },
...
}
```

What is the more appropriate and idiomatic way to handle this or are there other options? Also, in this case, what is the best way to authenticate/authorize that it is indeed the user?

**Comment by Kipitup** [comment]
Hello @Oyelowo,

I'm pretty much in the same spot as you. What did you end up doing?
Thanks

**Reply by Oyelowo** [reply]
I initially used next auth to call to my rust backend but didn't like the lack of predictability and flexibility especially for error handling. I found myself having to dig into next-auth implementation to be productive with it outside of the happy path. 
Thus, I decided to remove next-auth altogether and implement everything in rust which ended up smoothly. 

Rust backend handles all oauth and non-oauth interactions and stores the session in redis, then returns a 'set-cookie' header to the client. 

I have added support for GitHub and google and it's fairly easy to add new providers. Just implement a trait and that's it. 

I'm looking to open source it at some point. 

Hope this helps

**Reply by Manubi** [reply]
Hey! Did you open source it? As I am fairly new to rust it would be of great help! Thanks

**Reply by Oyelowo** [reply]
Hey! I’m open sourcing this along with an ultra-expressive ORM this month. Everything is done and ready but I’m working on finishing the book documentations.

**Reply by Manubi** [reply]
Cool! Where can I have a look at it? Don't want to miss it. :)
